### PR TITLE
revert HTTP return code change for user API, add tests

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Change HTTP return code back to 401 (from 403), as it is the case in 3.6 and before when calling `/_api/user/<name>` with invalid credentials.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9145/

### CODE REVIEW

File: arangod/GeneralServer/CommTask.cpp [hunk 1]
```diff
@@ -660,8 +660,6 @@ CommTask::Flow CommTask::canAccessPath(auth::TokenCache::Entry const& token,
         // `/_api/users/<name>` to check their passwords
         result = Flow::Continue;
         vc->forceReadOnly();
-      } else if (StringUtils::isPrefix(path, ApiUser)) {
-        result = Flow::Continue;
```
Comment (graetzer) [inline arangod/GeneralServer/CommTask.cpp]: I actually do not get why this is not equivalent ?

Comment (jsteemann) [reply arangod/GeneralServer/CommTask.cpp]: Probably what is missing there is the extra check for `vc->databaseAuthLevel() != auth::Level::NONE`

Comment (jsteemann) [reply arangod/GeneralServer/CommTask.cpp]: I debugged it and the difference is that the broken code sets the status from `Abort` to `Continue` via the now-deleted code:
```
       } else if (StringUtils::isPrefix(path, ApiUser)) {
         result = Flow::Continue;
       }
```
The correct behavior (as in 3.6 and the code of this PR) is to not make this change, and leave the status on `Abort` as decided here:
``` 
Flow result = userAuthenticated ? Flow::Continue : Flow::Abort;

// deny access to database with NONE
if (vc->databaseAuthLevel() == auth::Level::NONE && !StringUtils::isPrefix(path, ApiUser)) {
  result = Flow::Abort;
}
```
For the particular problem, the initial status is already `Abort`, and the correct code does not change it to anything else from there.
The broken code determined the initial status as `Abort`, but then overwrote it with `Continue` in the final `else if` that this PR deletes.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9151/

**Comment by jsteemann** [comment]
Tests blue

File: arangod/GeneralServer/CommTask.cpp (hunk 1 above)
Comment (graetzer) [reply arangod/GeneralServer/CommTask.cpp]: oh ok, so the problem is the user is not authenticated at all, why not just check for that later then ?
`else if (userAuthenticated && StringUtils::isPrefix(path, ApiUser))```

Comment (jsteemann) [reply arangod/GeneralServer/CommTask.cpp]: Ok, did.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9154/

File: arangod/GeneralServer/CommTask.cpp [hunk 2]
```diff
@@ -606,8 +606,9 @@ CommTask::Flow CommTask::canAccessPath(auth::TokenCache::Entry const& token,
 
   VocbaseContext* vc = static_cast<VocbaseContext*>(req.requestContext());
   TRI_ASSERT(vc != nullptr);
-  // deny access to database with NONE, (except /_api/user)
-  if (vc->databaseAuthLevel() == auth::Level::NONE/* && !StringUtils::isPrefix(path, ApiUser)*/) {
+  // deny access to database with NONE
+  if (result == Flow::Continue &&
```
Comment (graetzer) [inline arangod/GeneralServer/CommTask.cpp]: bit unnecessary to do this check then is it?

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9156/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/9159/

**Pull request merged**